sympy/tensor/array/arrayop.py

```diff
@@ -0,0 +1,142 @@
+import itertools
+
+import collections
+
+from sympy import S, Tuple, Matrix, prod
+
+from sympy.tensor.array import ImmutableDenseNDimArray
+from sympy.tensor.array.ndim_array import NDimArray
+
+
+def _arrayfy(a):
+    if isinstance(a, NDimArray):
+        return a
+    if isinstance(a, (Matrix, list, tuple, Tuple)):
+        return ImmutableDenseNDimArray(a)
+    return a
+
+
+def tensorproduct(*args):
+    """
+    Tensor product among scalars or array-like objects.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array import tensorproduct, Array
+    >>> from sympy.abc import x, y, z, t
+    >>> A = Array([[1, 2], [3, 4]])
+    >>> B = Array([x, y])
+    >>> tensorproduct(A, B)
+    [[[x, y], [2*x, 2*y]], [[3*x, 3*y], [4*x, 4*y]]]
```

Comment:
Printed code is in the _str_ method:
https://github.com/sympy/sympy/blob/ce0428d931fc1f785e4b323e0e1903417ebba64b/sympy/tensor/array/ndim_array.py#L224